domain/src/main/java/org/oppia/domain/exploration/ExplorationDataController.kt

```diff
@@ -0,0 +1,289 @@
+package org.oppia.domain.exploration
+
+import android.content.Context
+import androidx.lifecycle.LiveData
+import org.json.JSONArray
+import org.json.JSONException
+import javax.inject.Inject
+import javax.inject.Singleton
+import org.oppia.app.model.Exploration
+import org.oppia.util.data.AsyncResult
+import java.io.IOException
+import org.json.JSONObject
+import org.oppia.app.model.AnswerGroup
+import org.oppia.app.model.Outcome
+import org.oppia.app.model.Interaction
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.RuleSpec
+import org.oppia.app.model.State
+import org.oppia.app.model.StringList
+import org.oppia.app.model.SubtitledHtml
+import org.oppia.util.data.DataProviders
+
+const val TEST_EXPLORATION_ID_0 = "test_exp_id_0"
+const val TEST_EXPLORATION_ID_1 = "test_exp_id_1"
+private const val WELCOME_EXPLORATION_DATA_PROVIDER_ID = "WelcomeExplorationDataProvider"
+private const val ABBOUT_OPPIA_EXPLORATION_DATA_PROVIDER_ID = "AboutOppiaExplorationDataProvider"
+
+/** Controller for retrieving an exploration. */
+@Singleton
+class ExplorationDataController @Inject constructor(private val context: Context,
+                                                    private val dataProviders: DataProviders
+) {
+
+  private val welcomeExplorationDataProvider =
+    dataProviders.createInMemoryDataProviderAsync(
+      WELCOME_EXPLORATION_DATA_PROVIDER_ID, this::retrieveWelcomeExplorationAsync)
+  private val aboutOppiaExplorationDataProvider =
+    dataProviders.createInMemoryDataProviderAsync(
+      ABBOUT_OPPIA_EXPLORATION_DATA_PROVIDER_ID, this::retrieveAbboutOppiaExplorationAsync)
+
+  /**
+   * Returns an  [Exploration] given an ID.
+   */
+  fun getExplorationById(ID: String): LiveData<AsyncResult<Exploration>>? {
```

Comment:
We shouldn't return a nullable type here. Instead, if we can't fetch the requested ID the resulting LiveData should yield a failed AsyncResult. Otherwise we are sending mixed messages: the function cal fail in two different ways (no LiveData and a failing LiveData), which may be confusing to develop against.

Reply:
Also nit: the param should be ``id``

Reply:
Done